spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/GenericsHelpers.java

```diff
@@ -289,11 +317,27 @@ private static void assertEqualsUnsafe(Type type, Object expected, Object actual
     }
 
     switch (type.typeId()) {
+      case LONG:
+        assertThat(actual).as("Should be a long").isInstanceOf(Long.class);
+        if (expected instanceof Integer) {
+          assertThat(actual).as("Values didn't match").isEqualTo(((Number) expected).longValue());
+        } else {
+          assertThat(actual).as("Primitive value should be equal to expected").isEqualTo(expected);
+        }
+        break;
+      case DOUBLE:
+        assertThat(actual).as("Should be a double").isInstanceOf(Double.class);
+        if (expected instanceof Float) {
+          assertThat(Double.doubleToLongBits((double) actual))
+              .as("Values didn't match")
+              .isEqualTo(Double.doubleToLongBits(((Number) expected).doubleValue()));
+        } else {
+          assertThat(actual).as("Primitive value should be equal to expected").isEqualTo(expected);
+        }
+        break;
```

Comment:
This is needed because of the int to long, float to double type promotion tests. This is the same as what exists in TestHelpers#assertEqualsUnsafe

Reply:
I should double check why this wasn't failing before.

Reply:
Disregard, my first comment still stands.

I confused myself, we only  do `testReadsForTypePromotedColumns` for the vectorized parquet reader test so it makes sense that this wasn't failing before because before we were calling `TestHelpers#assertEqualsUnsafe` which has this same logic.